packages/caravan-wallets/src/jade.ts

```diff
@@ -0,0 +1,504 @@
+      return "sh(wsh(multi(k)))";
+    default:
+      throw new Error(`Unsupported addressType ${t}`);
+  }
+}
+
+function fingerprintFromHex(xfp: string): Uint8Array {
+  return Uint8Array.from(Buffer.from(xfp, "hex"));
+}
+
+function parseBip32Path(path_i: string): number[] {
+  let path = path_i;
+  if (path.startsWith("m/")) {
+    path = path.substring(2);
+  } else if (path.startsWith("m")) {
+    path = path.substring(1);
+    if (path.startsWith("/")) {
+      path = path.substring(1);
+    }
+  }
+  const segments = path.split("/");
+  const result: number[] = [];
+  for (const segment of segments) {
+    let hardened = false;
+    let numStr = segment;
+    if (segment.endsWith("'") || segment.endsWith("h")) {
+      hardened = true;
+      numStr = segment.slice(0, -1);
+    }
+    const index = parseInt(numStr, 10);
+    if (isNaN(index)) {
+      throw new Error(`Invalid path segment: ${segment}`);
+    }
+    result.push(index + (hardened ? 0x80000000 : 0));
+  }
+  return result;
+}
+
+function extractPathSuffix(
+  fullPathStr: string,
+  baseDerivation: number[],
+): number[] {
+  const fullPath = parseBip32Path(fullPathStr);
+  if (
+    fullPath.length < baseDerivation.length ||
+    !baseDerivation.every((v, i) => v === fullPath[i])
+  ) {
+    throw new Error(
+      `Path "${fullPathStr}" does not extend base derivation [${baseDerivation.join(",")}]`,
+    );
+  }
+  return fullPath.slice(baseDerivation.length);
+}
+
+function walletConfigToDescriptor(
+  cfg: MultisigWalletConfig,
+): MultisigDescriptor {
+  const signers: SignerDescriptor[] = cfg.extendedPublicKeys.map((ek) => ({
+    fingerprint: fingerprintFromHex(ek.xfp),
+    derivation: parseBip32Path(ek.bip32Path),
+    xpub: ek.xpub,
+    path: [],
+  }));
+
+  return {
+    variant: variantFromAddressType(cfg.addressType),
+    sorted: true,
+    threshold: cfg.quorum.requiredSigners,
+    signers,
+  };
+}
+
+export class JadeInteraction extends DirectKeystoreInteraction {
+  protected transport: JadeTransport;
+
+  protected ijade: IJadeInterface;
+
+  protected jade: IJade;
+
+  constructor() {
+    super();
+    this.transport = new SerialTransport({});
+    this.ijade = new JadeInterface(this.transport);
+    this.jade = new Jade(this.ijade);
+  }
+
+  /**
+   * Provides a list of status messages for the UI.
+   */
+  messages() {
+    const messages = super.messages();
+    messages.push({
+      state: PENDING,
+      level: INFO,
+      text: "Please connect your Jade device.",
+      code: "device.setup",
+    });
+    messages.push({
+      state: ACTIVE,
+      level: INFO,
+      text: "Communicating with Jade...",
+      code: "device.active",
+    });
+    return messages;
+  }
+
+  async withDevice<T>(
+    network: string,
+    f: (jade: IJade) => Promise<T>,
+  ): Promise<T> {
+    try {
+      await this.jade.connect();
+
+      const httpRequestFn = async (params: any): Promise<{ body: any }> => {
+        const url = params.urls[0];
```

Comment:
curious where this url gets sent from. is this authentication required for jade interactions? it seems somewhat concerning that there is some client-source url that is passed to an authentication endpoint call here. Are there any man in the middle risks we should be worried about or protecting against?

Reply:
bump. I get this might be required, but would like to understand this a little more. Maybe we could at least add some inline comments? I actually wonder if this is related to my issues trying to test out this MR? If I don't unlock with the app then this is what happens when trying to interact with the device. it just hangs here. If I do unlock with the app then I get the serial port conflict. 

<img width="1059" height="464" alt="Screenshot 2025-08-06 at 5 42 36 PM" src="https://github.com/user-attachments/assets/b9b0b4ca-d0cf-4d79-9892-d9d5c065f421" />